File: src/natural_language.rs
use std::sync::OnceLock;

use lindera::dictionary::load_dictionary;
use lindera::mode::Mode;
use lindera::segmenter::Segmenter;
use lindera::tokenizer::Tokenizer;
use lingua::Language::{
    Arabic, Bengali, Chinese, English, French, Hindi, Portuguese, Russian, Spanish, Urdu,
};
use lingua::{Language, LanguageDetector, LanguageDetectorBuilder};
use unicode_bidi::BidiInfo;
use unicode_normalization::UnicodeNormalization;
use unicode_segmentation::UnicodeSegmentation;

use crate::configuration::{LanguageIdentificationDetector, ParseConfiguration};
use crate::link_network::{LinkId, LinkMetadata, LinkNetwork, LinkType};
use crate::source::{ByteRange, Point, SourceSpan};

const LINGUA_LANGUAGES: [Language; 10] = [
    English, Chinese, Hindi, Spanish, Arabic, French, Bengali, Portuguese, Russian, Urdu,
];

#[derive(Clone, Debug, PartialEq, Eq)]
struct WordSegment {
    text: String,
    range: ByteRange,
}

pub fn annotate_natural_language(
    network: &mut LinkNetwork,
    document: LinkId,
    text: &str,
    language: &str,
    configuration: ParseConfiguration,
) {
    let Some(declared_language) = canonical_natural_language(language) else {
        return;
    };

    let document_span = span_for_range(text, 0, text.len());
    let detected_language =
        identify_language(text, configuration.language_identification_detector())
            .unwrap_or(declared_language);

    let region = network.insert_link(
        [document],
        LinkMetadata::new()
            .with_link_type(LinkType::Region)
            .with_named(true)
            .with_term("natural-language region")
            .with_language(declared_language)
            .with_span(document_span),
    );
    let language_link = network.insert_link(
        [region],
        LinkMetadata::new()
            .with_link_type(LinkType::Language)
            .with_named(true)
            .with_term(detected_language)
            .with_language(detected_language)
            .with_span(document_span),
    );

    insert_semantic_annotation(
        network,
        region,
        language_link,
        detector_term(configuration.language_identification_detector()),
        detected_language,
        document_span,
    );

    let (segmentation_engine, segments) = if declared_language == "Mandarin Chinese" {
        ("lindera-jieba", lindera_segments(text))
    } else {
        ("unicode-segmentation", unicode_segments(text))
    };

    insert_semantic_annotation(
        network,
        region,
        language_link,
        format!("segmentation:{segmentation_engine}"),
        detected_language,
        document_span,
    );

    for segment in segments {
        network.insert_link(
            [region],
            LinkMetadata::new()
                .with_link_type(LinkType::Token)
                .with_named(true)
                .with_term(segment.text)
                .with_language(detected_language)
                .with_span(span_for_range(
                    text,
                    segment.range.start(),
                    segment.range.end(),
                )),
        );
    }

    insert_unicode_annotations(
        network,
        region,
        language_link,
        text,
        detected_language,
        document_span,
    );
}

fn insert_unicode_annotations(
    network: &mut LinkNetwork,
    region: LinkId,
    language_link: LinkId,
    text: &str,
    language: &str,
    span: SourceSpan,
) {
    let nfc = text.nfc().collect::<String>();
    let nfd = text.nfd().collect::<String>();

    insert_semantic_annotation(
        network,
        region,
        language_link,
        format!("normalization:nfc:{}", normalization_status(text, &nfc)),
        language,
        span,
    );
    insert_semantic_annotation(
        network,
        region,
        language_link,
        format!("normalization:nfd:{}", normalization_status(text, &nfd)),
        language,
        span,
    );
    insert_semantic_annotation(
        network,
        region,
        language_link,
        format!("bidi:{}", bidi_direction(text)),
        language,
        span,
    );
}

fn insert_semantic_annotation(
    network: &mut LinkNetwork,
    region: LinkId,
    language_link: LinkId,
    term: impl Into<String>,
    language: &str,
    span: SourceSpan,
) -> LinkId {
    network.insert_link(
        [region, language_link],
        LinkMetadata::new()
            .with_link_type(LinkType::Semantic)
            .with_named(true)
            .with_term(term)
            .with_language(language)
            .with_span(span),
    )
}

const fn detector_term(detector: LanguageIdentificationDetector) -> &'static str {
    match detector {
        LanguageIdentificationDetector::Lingua => "identifier:lingua",
        LanguageIdentificationDetector::Whatlang => "identifier:whatlang",
    }
}

fn identify_language(text: &str, detector: LanguageIdentificationDetector) -> Option<&'static str> {
    match detector {
        LanguageIdentificationDetector::Lingua => identify_with_lingua(text),
        LanguageIdentificationDetector::Whatlang => identify_with_whatlang(text),
    }
}

fn identify_with_lingua(text: &str) -> Option<&'static str> {
    static DETECTOR: OnceLock<LanguageDetector> = OnceLock::new();
    let detector =
        DETECTOR.get_or_init(|| LanguageDetectorBuilder::from_languages(&LINGUA_LANGUAGES).build());
    detector
        .detect_language_of(text)
        .map(canonical_lingua_language)
}

fn identify_with_whatlang(text: &str) -> Option<&'static str> {
    canonical_whatlang_language(whatlang::detect(text)?.lang())
}

const fn canonical_lingua_language(language: Language) -> &'static str {
    match language {
        English => "English",
        Chinese => "Mandarin Chinese",
        Hindi => "Hindi",
        Spanish => "Spanish",
        Arabic => "Modern Standard Arabic",
        French => "French",
        Bengali => "Bengali",
        Portuguese => "Portuguese",
        Russian => "Russian",
        Urdu => "Urdu",
    }
}

const fn canonical_whatlang_language(language: whatlang::Lang) -> Option<&'static str> {
    match language {
        whatlang::Lang::Eng => Some("English"),
        whatlang::Lang::Cmn => Some("Mandarin Chinese"),
        whatlang::Lang::Hin => Some("Hindi"),
        whatlang::Lang::Spa => Some("Spanish"),
        whatlang::Lang::Ara => Some("Modern Standard Arabic"),
        whatlang::Lang::Fra => Some("French"),
        whatlang::Lang::Ben => Some("Bengali"),
        whatlang::Lang::Por => Some("Portuguese"),
        whatlang::Lang::Rus => Some("Russian"),
        whatlang::Lang::Urd => Some("Urdu"),
        _ => None,
    }
}

fn canonical_natural_language(language: &str) -> Option<&'static str> {
    match language.to_ascii_lowercase().as_str() {
        "english" => Some("English"),
        "mandarin" | "mandarin chinese" | "chinese" => Some("Mandarin Chinese"),
        "hindi" => Some("Hindi"),
        "spanish" => Some("Spanish"),
        "arabic" | "modern standard arabic" => Some("Modern Standard Arabic"),
        "french" => Some("French"),
        "bengali" => Some("Bengali"),
        "portuguese" => Some("Portuguese"),
        "russian" => Some("Russian"),
        "urdu" => Some("Urdu"),
        _ => None,
    }
}

fn lindera_segments(text: &str) -> Vec<WordSegment> {
    let Ok(dictionary) = load_dictionary("embedded://jieba") else {
        return Vec::new();
    };
    let segmenter = Segmenter::new(Mode::Normal, dictionary, None);
    let tokenizer = Tokenizer::new(segmenter);
    let Ok(tokens) = tokenizer.tokenize(text) else {
        return Vec::new();
    };

    tokens
        .into_iter()
        .filter(|token| contains_word_character(token.surface.as_ref()))
        .map(|token| WordSegment {
            text: token.surface.into_owned(),
            range: ByteRange::new(token.byte_start, token.byte_end),
        })
        .collect()
}

fn unicode_segments(text: &str) -> Vec<WordSegment> {
    text.split_word_bound_indices()
        .filter(|(_start, segment)| contains_word_character(segment))
        .map(|(start, segment)| WordSegment {
            text: segment.to_string(),
            range: ByteRange::new(start, start + segment.len()),
        })
        .collect()
}

fn contains_word_character(text: &str) -> bool {
    text.chars().any(char::is_alphanumeric)
}

fn normalization_status(original: &str, normalized: &str) -> &'static str {
    if original == normalized {
        "stable"
    } else {
        "changes"
    }
}

fn bidi_direction(text: &str) -> &'static str {
    let bidi = BidiInfo::new(text, None);
    if bidi
        .paragraphs
        .iter()
        .any(|paragraph| paragraph.level.is_rtl())
    {
        "rtl"
    } else {
        "ltr"
    }
}

fn span_for_range(text: &str, start: usize, end: usize) -> SourceSpan {
    SourceSpan::new(
        ByteRange::new(start, end),
        point_at_byte(text, start),
        point_at_byte(text, end),
    )
}

fn point_at_byte(text: &str, byte: usize) -> Point {
    let mut row = 0;
    let mut column = 0;

    for (index, character) in text.char_indices() {
        if index >= byte {
            break;
        }
        if character == '\n' {
            row += 1;
            column = 0;
        } else {
            column += 1;
        }
    }

    Point::new(row, column)
}

File: js/src/natural-language.js
import {
  ByteRange,
  LinkFlags,
  LinkMetadata,
  LinkType,
  Point,
  SourceSpan,
} from './primitives.js';
import { seedStatehoodWorkedExample } from './concept-ontology.js';
import { LANGUAGE_IDENTIFIER_TERM, identifyLanguage } from './language-identification.js';
import { BIDI_RANGES } from './unicode-bidi-classes.js';

const encoder = new TextEncoder();

const CANONICAL_NATURAL_LANGUAGES = new Map([
  ['english', 'English'],
  ['en', 'English'],
  ['mandarin', 'Mandarin Chinese'],
  ['mandarin chinese', 'Mandarin Chinese'],
  ['chinese', 'Mandarin Chinese'],
  ['zh', 'Mandarin Chinese'],
  ['hindi', 'Hindi'],
  ['hi', 'Hindi'],
  ['spanish', 'Spanish'],
  ['es', 'Spanish'],
  ['arabic', 'Modern Standard Arabic'],
  ['modern standard arabic', 'Modern Standard Arabic'],
  ['ar', 'Modern Standard Arabic'],
  ['french', 'French'],
  ['fr', 'French'],
  ['bengali', 'Bengali'],
  ['bn', 'Bengali'],
  ['portuguese', 'Portuguese'],
  ['pt', 'Portuguese'],
  ['russian', 'Russian'],
  ['ru', 'Russian'],
  ['urdu', 'Urdu'],
  ['ur', 'Urdu'],
]);

// BCP 47 locales handed to Intl.Segmenter so dictionary-based word breaking
// applies where the language needs it (Mandarin has no spaces between words).
const SEGMENTER_LOCALES = new Map([
  ['English', 'en'],
  ['Mandarin Chinese', 'zh'],
  ['Hindi', 'hi'],
  ['Spanish', 'es'],
  ['Modern Standard Arabic', 'ar'],
  ['French', 'fr'],
  ['Bengali', 'bn'],
  ['Portuguese', 'pt'],
  ['Russian', 'ru'],
  ['Urdu', 'ur'],
]);

const STARTER_GRAMMAR_PROVENANCE =
  'repo-authored starter pass/fail sentence; license: Unlicense; ' +
  'morphosyntax tag names use Universal Dependencies v2 UPOS/UFeats/deprel vocabulary; ' +
  'no UD treebank sentence data imported';

/** Executable natural-language grammar fixture pairs for each target language. */
export const NATURAL_LANGUAGE_GRAMMAR_FIXTURES = Object.freeze([
  ['English', 'Hawaii is a state.\n', 'Hawaii are a state.\n'],
  ['Mandarin Chinese', '你好。\n', '你好 的。\n'],
  ['Hindi', 'नमस्ते।\n', 'नमस्ते है।\n'],
  ['Spanish', 'Hawaii es un estado.\n', 'Hawaii son un estado.\n'],
  ['Modern Standard Arabic', 'مرحبا.\n', 'مرحبا هو.\n'],
  ['French', 'Hawaii est un etat.\n', 'Hawaii sont un etat.\n'],
  ['Bengali', 'নমস্কার।\n', 'নমস্কার আছে।\n'],
  ['Portuguese', 'Hawaii e um estado.\n', 'Hawaii sao um estado.\n'],
  ['Russian', 'Гавайи это штат.\n', 'Гавайи это штаты.\n'],
  ['Urdu', 'سلام۔\n', 'سلام ہے۔\n'],
].map(([language, grammaticalSource, ungrammaticalSource]) => Object.freeze({
  language,
  grammaticalSource,
  ungrammaticalSource,
  provenance: STARTER_GRAMMAR_PROVENANCE,
})));

const NO_FEATURES = [];
const NUMBER_SING = ['Number=Sing'];
const NUMBER_PLUR = ['Number=Plur'];
const INDEFINITE_ARTICLE = ['Definite=Ind', 'PronType=Art'];
const AUX_SINGULAR = ['Mood=Ind', 'Number=Sing', 'Person=3', 'Tense=Pres', 'VerbForm=Fin'];
const AUX_PLURAL = ['Mood=Ind', 'Number=Plur', 'Person=3', 'Tense=Pres', 'VerbForm=Fin'];
const PRON_SINGULAR = ['Number=Sing', 'Person=3', 'PronType=Prs'];

const PUNCT_ANALYSIS = { upos: 'PUNCT', features: NO_FEATURES, deprel: 'punct' };

const SENTENCE_GRAMMARS = new Map([
  ['English', ['Hawaii', 'is', 'a', 'state', '.']],
  ['Mandarin Chinese', ['你好', '。']],
  ['Hindi', ['नमस्ते', '।']],
  ['Spanish', ['Hawaii', 'es', 'un', 'estado', '.']],
  ['Modern Standard Arabic', ['مرحبا', '.']],
  ['French', ['Hawaii', 'est', 'un', 'etat', '.']],
  ['Bengali', ['নমস্কার', '।']],
  ['Portuguese', ['Hawaii', 'e', 'um', 'estado', '.']],
  ['Russian', ['Гавайи', 'это', 'штат', '.']],
  ['Urdu', ['سلام', '۔']],
]);

const LEXICON = [
  ['English', 'Hawaii', 'PROPN', NUMBER_SING, 'nsubj'],
  ['English', 'is', 'AUX', AUX_SINGULAR, 'cop'],
  ['English', 'are', 'AUX', AUX_PLURAL, 'cop'],
  ['English', 'a', 'DET', INDEFINITE_ARTICLE, 'det'],
  ['English', 'state', 'NOUN', NUMBER_SING, 'root'],
  ['Mandarin Chinese', '你好', 'INTJ', NO_FEATURES, 'root'],
  ['Mandarin Chinese', '的', 'PART', NO_FEATURES, 'mark'],
  ['Hindi', 'नमस्ते', 'INTJ', NO_FEATURES, 'root'],
  ['Hindi', 'है', 'AUX', AUX_SINGULAR, 'cop'],
  ['Spanish', 'Hawaii', 'PROPN', NUMBER_SING, 'nsubj'],
  ['Spanish', 'es', 'AUX', AUX_SINGULAR, 'cop'],
  ['Spanish', 'son', 'AUX', AUX_PLURAL, 'cop'],
  ['Spanish', 'un', 'DET', INDEFINITE_ARTICLE, 'det'],
  ['Spanish', 'estado', 'NOUN', NUMBER_SING, 'root'],
  ['Modern Standard Arabic', 'مرحبا', 'INTJ', NO_FEATURES, 'root'],
  ['Modern Standard Arabic', 'هو', 'PRON', PRON_SINGULAR, 'dep'],
  ['French', 'Hawaii', 'PROPN', NUMBER_SING, 'nsubj'],
  ['French', 'est', 'AUX', AUX_SINGULAR, 'cop'],
  ['French', 'sont', 'AUX', AUX_PLURAL, 'cop'],
  ['French', 'un', 'DET', INDEFINITE_ARTICLE, 'det'],
  ['French', 'etat', 'NOUN', NUMBER_SING, 'root'],
  ['Bengali', 'নমস্কার', 'INTJ', NO_FEATURES, 'root'],
  ['Bengali', 'আছে', 'AUX', AUX_SINGULAR, 'cop'],
  ['Portuguese', 'Hawaii', 'PROPN', NUMBER_SING, 'nsubj'],
  ['Portuguese', 'e', 'AUX', AUX_SINGULAR, 'cop'],
  ['Portuguese', 'sao', 'AUX', AUX_PLURAL, 'cop'],
  ['Portuguese', 'um', 'DET', INDEFINITE_ARTICLE, 'det'],
  ['Portuguese', 'estado', 'NOUN', NUMBER_SING, 'root'],
  ['Russian', 'Гавайи', 'PROPN', NUMBER_SING, 'nsubj'],
  ['Russian', 'это', 'PRON', PRON_SINGULAR, 'cop'],
  ['Russian', 'штат', 'NOUN', NUMBER_SING, 'root'],
  ['Russian', 'штаты', 'NOUN', NUMBER_PLUR, 'root'],
  ['Urdu', 'سلام', 'INTJ', NO_FEATURES, 'root'],
  ['Urdu', 'ہے', 'AUX', AUX_SINGULAR, 'cop'],
].map(([language, surface, upos, features, deprel]) => ({
  language,
  surface,
  analysis: { upos, features, deprel },
}));

const SENTENCE_PUNCTUATION = new Set(['.', '!', '?', '।', '。', '۔']);

/** Returns the canonical name for a supported natural-language name or alias. */
export function canonicalNaturalLanguage(language) {
  return CANONICAL_NATURAL_LANGUAGES.get(String(language).toLowerCase());
}

/**
 * Adds the natural-language layer under a parsed document: a region with its
 * identified language, word segments, Universal Dependencies morphosyntax,
 * Unicode normalization and bidi annotations.
 */
export function annotateNaturalLanguage(network, document, text, language) {
  const declaredLanguage = canonicalNaturalLanguage(language);
  if (!declaredLanguage) return;

  const lines = new ByteLineIndex(text);
  const documentSpan = lines.span(0, encoder.encode(text).length);
  const detectedLanguage = identifyLanguage(text) ?? declaredLanguage;

  const languageLink = network.insertLink(
    [],
    LinkMetadata.new()
      .withLinkType(LinkType.Language)
      .withNamed(true)
      .withTerm(detectedLanguage)
      .withLanguage(detectedLanguage)
      .withSpan(documentSpan),
  );
  const annotations = [
    insertSemanticAnnotation(network, languageLink, LANGUAGE_IDENTIFIER_TERM, detectedLanguage, documentSpan),
    insertSemanticAnnotation(
      network,
      languageLink,
      'segmentation:intl-segmenter',
      detectedLanguage,
      documentSpan,
    ),
  ];

  // Segments belong to the declared region; the identifier's verdict lives on
  // the Language link and its annotation.
  const segments = wordSegments(text, declaredLanguage).map(({ segment, start, end }) =>
    network.insertLink(
      [],
      LinkMetadata.new()
        .withLinkType(LinkType.SourceToken)
        .withNamed(true)
        .withTerm(segment)
        .withLanguage(declaredLanguage)
        .withSpan(lines.span(start, end)),
    ));

  const sentence = annotateMorphosyntax(network, text, lines, declaredLanguage, documentSpan);
  annotations.push(
    ...unicodeAnnotations(text).map((term) =>
      insertSemanticAnnotation(network, languageLink, term, detectedLanguage, documentSpan)),
  );
  const workedExample = insertWorkedExampleAnnotation(network, text, declaredLanguage, documentSpan);
  if (workedExample) annotations.push(workedExample);

  const region = network.insertLink(
    [document, languageLink, ...(sentence ? [sentence] : []), ...segments, ...annotations],
    LinkMetadata.new()
      .withLinkType(LinkType.Region)
      .withNamed(true)
      .withTerm('natural-language region')
      .withLanguage(declaredLanguage)
      .withSpan(documentSpan),
  );
  return region;
}

function insertSemanticAnnotation(network, languageLink, term, language, span) {
  return network.insertLink(
    [languageLink],
    LinkMetadata.new()
      .withLinkType(LinkType.Semantic)
      .withNamed(true)
      .withTerm(term)
      .withLanguage(language)
      .withSpan(span),
  );
}

function insertWorkedExampleAnnotation(network, text, language, span) {
  if (!isStatehoodWorkedExample(text, language)) return undefined;
  const concepts = seedStatehoodWorkedExample(network);
  return network.insertLink(
    [concepts.proposition, concepts.subject, concepts.object],
    LinkMetadata.new()
      .withLinkType(LinkType.Semantic)
      .withNamed(true)
      .withTerm('proposition:statehood')
      .withLanguage(language)
      .withSpan(span),
  );
}

function isStatehoodWorkedExample(text, language) {
  switch (language) {
    case 'English':
      return text.trim() === 'Hawaii is a state.';
    case 'Russian':
      return text.trim() === 'Гавайи это штат.';
    default:
      return false;
  }
}

function wordSegments(text, language) {
  const segmenter = new Intl.Segmenter(SEGMENTER_LOCALES.get(language), { granularity: 'word' });
  const segments = [];
  let byte = 0;
  let index = 0;
  for (const { segment, index: segmentIndex } of segmenter.segment(text)) {
    byte += encoder.encode(text.slice(index, segmentIndex)).length;
    index = segmentIndex;
    const length = encoder.encode(segment).length;
    if (/[\p{Alphabetic}\p{N}]/u.test(segment)) {
      segments.push({ segment, start: byte, end: byte + length });
    }
  }
  return segments;
}

function unicodeAnnotations(text) {
  return [
    `normalization:nfc:${text.normalize('NFC') === text ? 'stable' : 'changes'}`,
    `normalization:nfd:${text.normalize('NFD') === text ? 'stable' : 'changes'}`,
    `bidi:${bidiDirection(text)}`,
  ];
}

/** Adds the sentence, form, UPOS, UFeats and deprel links; returns the sentence. */
function annotateMorphosyntax(network, text, lines, language, span) {
  const tokens = grammarTokens(text);
  if (tokens.length === 0) return undefined;

  const accepted = SENTENCE_GRAMMARS.get(language);
  const sentenceIsAccepted = accepted !== undefined &&
    tokens.length === accepted.length &&
    tokens.every((token, index) => token.text === accepted[index]);
  const shouldReportErrors = isRegisteredGrammarFixture(language, text) && !sentenceIsAccepted;

  const children = [];
  for (const token of tokens) {
    const tokenSpan = lines.span(token.start, token.end);
    const analysis = morphologyFor(language, token.text);
    const formChildren = analysis
      ? [
        insertSyntax(network, `upos:${analysis.upos}`, [], language, tokenSpan),
        ...analysis.features.map((feature) =>
          insertSyntax(network, `ufeat:${feature}`, [], language, tokenSpan)),
      ]
      : [];
    const form = insertSyntax(network, `form:${token.text}`, formChildren, language, tokenSpan);
    children.push(form);
    if (analysis) {
      children.push(insertSyntax(network, `deprel:${analysis.deprel}`, [form], language, tokenSpan));
    } else if (shouldReportErrors) {
      children.push(insertSyntax(
        network,
        'natural-language:error:unknown-token',
        [form],
        language,
        tokenSpan,
        LinkFlags.error(),
      ));
    }
  }
  if (shouldReportErrors) {
    children.push(insertSyntax(
      network,
      'natural-language:error:grammar',
      [],
      language,
      span,
      LinkFlags.error(),
    ));
  }

  // Inserted last so the sentence is the parent of every form it lists, even
  // though deprel and error links also reference the forms they relate.
  return insertSyntax(
    network,
    'natural-language:sentence',
    children,
    language,
    span,
    shouldReportErrors ? LinkFlags.containingError() : LinkFlags.clean(),
  );
}

function insertSyntax(network, term, children, language, span, flags = LinkFlags.clean()) {
  return network.insertSyntaxNode(language, term, children, { span, flags });
}

function morphologyFor(language, surface) {
  if (SENTENCE_PUNCTUATION.has(surface)) return PUNCT_ANALYSIS;
  return LEXICON.find((entry) => entry.language === language && entry.surface === surface)
    ?.analysis;
}

function isRegisteredGrammarFixture(language, text) {
  return NATURAL_LANGUAGE_GRAMMAR_FIXTURES.some((fixture) =>
    fixture.language === language &&
    (text === fixture.grammaticalSource || text === fixture.ungrammaticalSource));
}

function grammarTokens(text) {
  const tokens = [];
  let wordStart;
  let byte = 0;
  const pushWord = (end) => {
    if (wordStart === undefined) return;
    if (wordStart.byte !== end) {
      tokens.push({ text: text.slice(wordStart.index, wordStart.index + wordStart.length), start: wordStart.byte, end });
    }
    wordStart = undefined;
  };
  let index = 0;
  for (const character of text) {
    const length = encoder.encode(character).length;
    if (/^\p{White_Space}$/u.test(character)) {
      pushWord(byte);
    } else if (SENTENCE_PUNCTUATION.has(character)) {
      pushWord(byte);
      tokens.push({ text: character, start: byte, end: byte + length });
    } else if (wordStart === undefined) {
      wordStart = { byte, index, length: 0 };
    }
    if (wordStart !== undefined) wordStart.length = index + character.length - wordStart.index;
    byte += length;
    index += character.length;
  }
  pushWord(byte);
  return tokens;
}

/**
 * Paragraph direction per UAX #9 rules P1-P3, matching unicode-bidi's
 * BidiInfo: the text is right-to-left when any paragraph's first strong
 * character outside isolates is R or AL.
 */
export function bidiDirection(text) {
  let paragraphLevel;
  let isolates = 0;
  let rtl = false;
  for (const character of text) {
    const kind = bidiKind(character.codePointAt(0));
    if (kind === 'B') {
      rtl ||= paragraphLevel === 'rtl';
      paragraphLevel = undefined;
      isolates = 0;
    } else if (kind === 'L' || kind === 'R') {
      if (isolates === 0 && paragraphLevel === undefined) {
        paragraphLevel = kind === 'R' ? 'rtl' : 'ltr';
      }
    } else if (kind === 'I') {
      isolates += 1;
    } else if (kind === 'P') {
      isolates = Math.max(0, isolates - 1);
    }
  }
  return rtl || paragraphLevel === 'rtl' ? 'rtl' : 'ltr';
}

let bidiRanges;

function bidiKind(codePoint) {
  bidiRanges ??= BIDI_RANGES.split(' ').map((entry) => {
    const [start, end, kind] = entry.split('.');
    return [Number.parseInt(start, 36), Number.parseInt(end, 36), kind];
  });
  let low = 0;
  let high = bidiRanges.length - 1;
  while (low <= high) {
    const middle = (low + high) >> 1;
    const [start, end, kind] = bidiRanges[middle];
    if (codePoint < start) high = middle - 1;
    else if (codePoint > end) low = middle + 1;
    else return kind;
  }
  return 'L';
}

class ByteLineIndex {
  constructor(text) {
    this.starts = [0];
    let byte = 0;
    for (const character of text) {
      byte += encoder.encode(character).length;
      if (character === '\n') this.starts.push(byte);
    }
  }

  point(byte) {
    let low = 0;
    let high = this.starts.length - 1;
    while (low < high) {
      const middle = (low + high + 1) >> 1;
      if (this.starts[middle] <= byte) low = middle;
      else high = middle - 1;
    }
    return new Point(low, byte - this.starts[low]);
  }

  span(start, end) {
    return new SourceSpan(new ByteRange(start, end), this.point(start), this.point(end));
  }
}
